packages/adapter/src/lib/_Types.ts

```diff
@@ -0,0 +1,402 @@
+import type { ChangeFileFunction } from '@iobroker/db-objects-redis';
+import type { IdObject } from './adapter/utils';
+
+export interface AdapterOptions {
+    subscribesChange?: (subs: Record<string, { regex: RegExp }>) => void;
+    /** If the adapter collects logs from all adapters (experts only). Default: false */
+    logTransporter?: boolean;
+    /** if true, the date format from system.config */
+    useFormatDate?: boolean;
+    /** if it is possible for other instances to retrive states of this adapter automatically */
+    subscribable?: boolean;
+    /** compact group instance if running in compact mode */
+    compactInstance?: number;
+    /** if desired to have oStates. This is a list with all states values, and it will be updated automatically. */
+    states?: boolean;
+    /** if desired to have oObjects. This is a list with all states, channels and devices of this adapter, and it will be updated automatically.*/
+    objects?: boolean;
+    /** instance number of adapter */
+    instance?: number;
+    /** adapter directory name */
+    dirname?: string;
+    /** flag which defaults to true - if true, adapter warns if states are set without a corresponding existing object */
+    strictObjectChecks?: boolean;
+    /** If true runs in compact mode */
+    compact?: boolean;
+    /** configuration of the connection to controller */
+    config?: AdapterOptionsConfig;
+    /** name of the adapter. Must be exactly the same as directory name. */
+    name: string;
+    /** If true, the systemConfig (iobroker.json) will be available in this.systemConfig */
+    systemConfig?: boolean;
+    /** callback function (id, obj) that will be called if object changed */
+    objectChange?: ioBroker.ObjectChangeHandler;
+    /** callback function (id, obj) that will be called if state changed */
+    stateChange?: ioBroker.StateChangeHandler;
+    /** callback function (id, file) that will be called if file changed */
+    fileChange?: ChangeFileFunction;
+    /** callback to inform about new message the adapter */
+    message?: ioBroker.MessageHandler;
+    /** callback to stop the adapter */
+    unload?: ioBroker.UnloadHandler;
+    /** called when adapter is ready */
+    ready?: ioBroker.ReadyHandler;
+    /** called on reconnection to DB */
+    reconnect?: () => VoidLikeCallback;
+    /** Handler to handle uncaught exceptions, return true if no further handling required */
+    error?: ioBroker.ErrorHandler;
+}
+
+export type AdapterConfig = AdapterOptions | ioBroker.InstanceObject;
```

Comment:
These are two very different things, does that make sense to union them under one type?

Reply:
Unfortunately a weird code behavior leads to adapter config being one of these two things, but I can also make the union type just on the assignment

Reply:
I think that would be less confusing/prone to reusing this weird union elsewhere.